### PR TITLE
Move the 'GitRepositoryProvider' to compare urls using their canonical representation, correctly accepting results where repositories only differ by '.git'

### PR DESCRIPTION
**Pull request by francescomikulis**
### Motivation:

Resolving local package dependencies often outputs a warning:
<path> is not valid git repository for '<repo>', will fetch again.

### Modifications:

The underlying issue is that updating packages with a local SCM path dependency always fail, as the shell-based repository URL has a file:// scheme, while the absolute path string does not.

### Result:

After this change we only compare the canonical URLs, which will resolve the inconsistency of comparing strings without considering the scheme and / or path extension.

Validating the origin of checked-out repositories no longer ensures that the path extension (often .git) matches the repository's specifier, as different git clients inconsistently preserve the path extension of the remote. Moving to the CanonicalPackageURL also ensures that absolute paths are treated as urls with a file:// scheme, matching git's behavior.

### CODE REVIEW

**Comment by rauhul** [comment]
Im gonna pre-flight what I guessing @MaxDesiatov will say: "Could you add some tests for this?"

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
Yeah, we definitely need tests for this before merging.

**Comment by neonichu** [comment]
The motivation for local dependencies makes sense to me, but I don't think it's correct to ignore schemes and path extensions here, the whole purpose of the validation was to make sure that we have the "correct" repo. If we're now not doing that anymore, it might make more sense to remove the entire validation logic.

**Comment by bnbarham** [comment]
> If we're now not doing that anymore, it might make more sense to remove the entire validation logic.

TBH I'd be fine with this - the repository cache seems like the correct layer to validate this. Otherwise we're just going to run into issues where the cache says the repositories match, but this validation disagrees.

File: Sources/SourceControl/GitRepository.swift [hunk 1]
```diff
@@ -207,12 +210,21 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
...
     public func isValidDirectory(_ directory: Basics.AbsolutePath, for repository: RepositorySpecifier) throws -> Bool {
         let remoteURL = try self.git.run(["-C", directory.pathString, "config", "--get", "remote.origin.url"])
-        return remoteURL == repository.url
+        switch repository.location {
+        case .url(let url):
+            // Compare the canonical representation, which will drop any suffix and canonicalize scp-style urls
+            return CanonicalPackageURL(remoteURL) == CanonicalPackageURL(url.absoluteString)
+        case .path(let absolutePath):
+            // Compare the canonical representation, which will drop any suffix
+            return CanonicalPackageLocation(remoteURL) == CanonicalPackageLocation(absolutePath.pathString)
```
Comment (bnbarham) [inline Sources/SourceControl/GitRepository.swift:226]: I doubt this matters in practice, but the [caching logic](https://github.com/swiftlang/swift-package-manager/blob/main/Sources/SourceControl/RepositoryManager.swift#L617) always uses URL and it's probably better to do that here as well. That wouldn't fix the case mentioned in the description, but to fix that you could update `computeCanonicalLocation` to return a `file` scheme if one wasn't given (which seems somewhat reasonable given we're already inferring `ssh` if the user is `git`). Does that seem fine to you @neonichu?

**Comment by MaxDesiatov** [comment]
@swift-ci test

File: Sources/SourceControl/GitRepository.swift [hunk 2]
```diff
@@ -207,12 +210,21 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
 
     public func isValidDirectory(_ directory: Basics.AbsolutePath) throws -> Bool {
         let result = try self.git.run(["-C", directory.pathString, "rev-parse", "--git-dir"])
```
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SourceControl/GitRepository.swift:212]: @francescomikulis pointed out that this actually calls `rev-parse` rather than `config --get remote.origin.url`. That is, it's just checking whether this is a git directory at all (rather than checking that the remote matches). Ignore me!

File: Sources/SourceControl/GitRepository.swift [hunk 3]
```diff
@@ -207,12 +210,14 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
...
-        return result == ".git" || result == "." || result == directory.pathString
```
Comment (bnbarham) [inline Sources/SourceControl/GitRepository.swift:210]: Actually, why are we even checking the resulting string at all? Doesn't `rev-parse` just fail if it isn't a git directory and that's all we care about in this function?

Comment (francescomikulis) [reply Sources/SourceControl/GitRepository.swift:210]: Running it on a non-directory yields the following in the terminal:
fatal: not a git repository (or any of the parent directories): .git

Comment (bnbarham) [reply Sources/SourceControl/GitRepository.swift:210]: Hmm, so it returns:
```
       --git-dir
           Show $GIT_DIR if defined. Otherwise show the path to the .git directory. The path shown, when relative, is
           relative to the current working directory.

           If $GIT_DIR is not defined and the current directory is not detected to lie in a Git repository or work tree
           print a message to stderr and exit with nonzero status.
```

I'm not sure if this function is purposefully checking that $GIT_DIR isn't set or not?

I would be fine with you leaving this TBH, it's unrelated to the change you actually care about.

File: Sources/SourceControl/Repository.swift [hunk 4]
```diff
@@ -32,18 +33,18 @@ public struct RepositorySpecifier: Hashable, Sendable {
     }
 
     /// The location of the repository as string.
-    public var url: String {
+    public var url: CanonicalPackageURL {
         switch self.location {
-        case .path(let path): return path.pathString
-        case .url(let url): return url.absoluteString
+        case .path(let path): return CanonicalPackageURL(path.pathString)
+        case .url(let url): return CanonicalPackageURL(url.absoluteString)
```
Comment (bnbarham) [inline Sources/SourceControl/Repository.swift:39]: Why this change? The canonical URL isn't necessarily the actual URL 🤔

Comment (francescomikulis) [reply Sources/SourceControl/Repository.swift]: I removed this, it wasn't necessary.

File: Tests/WorkspaceTests/SourceControlPackageContainerTests.swift [hunk 5]
```diff
@@ -89,74 +89,6 @@ private class MockRepository: Repository {
     }
 }
 
-private class MockRepositories: RepositoryProvider {
```
Comment (francescomikulis) [inline Tests/WorkspaceTests/SourceControlPackageContainerTests.swift:92]: This was unused.

**Review by bnbarham — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by MaxDesiatov — COMMENTED** [review]
IMO the change seems noteworthy enough to be included in `CHANGELOG.md`?

**Review by Hesham-atta — APPROVED** [review]

**Comment by robinkunde** [comment]
I was preparing my own PRs for this issue, but this one is much further along. Great work @francescomikulis!

One question I had was: Would it make sense to canonicalize the origin URL before populating the cache? As it stands now, whether the cached repo uses `https://github.com/swiftlang/swift-package-manager.git` or `https://github.com/swiftlang/swift-package-manager` for example depends on which version of the URL was encountered first in a dependency graph. This PR fixes the lookup logic, but I think the logic for populating the cache should be as predictable as possible too.

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test Windows platform
